### PR TITLE
Mission Control with partial Unit Tests

### LINKED ISSUE #3: Feat All

### PR DESCRIPTION
**Pull request by Philipp66904**
This is a first implementation for the Mission Control node that can be used for further tests.
Unit tests will follow later.

### CODE REVIEW

**Comment by Philipp66904** [comment]
Note: This does not build, as it requires Messages from interfaces that are currently not merged into dev

**Comment by J0hn-Lawrence** [comment]
It will take a while to review all of the >5k LoC :laughing:, please bear with me  @Philipp66904 I am working on it!

**Comment by Philipp66904** [comment]
Luckily, almost 3k of it is an auto generated and adapted Doxyfile 😉

File: assets/mission_test.json
Comment (J0hn-Lawrence) [inline assets/mission_test.json:1]: Nice touch using the coordinates for the flight area already! But the geofence is not yet confirmed and currently just a guesstimate by you, right?

File: include/mission_control.hpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+   private:
+    // General
+    MissionState_t mission_state = prepare_mission;  //!< Main mission state
+    std::string active_marker_name =
+        "init";  //!< Name of the marker that is currently active
+    bool job_finished_successfully =
+        false;  //!< Will be set to true after a successfull job_finished
+                //!< message was received from the active node
+    nlohmann::json
+        job_finished_payload;  //!< If a job finished successfully, the payload
```
Comment (J0hn-Lawrence) [inline include/mission_control.hpp:70]: this is always the last finished jobs payload? (Asking as there are multiple finished jobs expected for a complete mission)

File: include/mission_control.hpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        true;  //!< If set to true, the first event loop after a mission_state
+               //!< change is happening. Will be set to false when calling
+               //!< get_state_first_loop().
+    std::map<std::string, ros_node>
+        node_map;  //!< Has an entry for every ros node
+    std::string active_node_id =
+        "";  //!< node_id that is currently allowed to send data to the FCC
+             //!< interface, set to "" if none is allowed
+
+    size_t current_command_id = 0;
```
Comment (J0hn-Lawrence) [inline include/mission_control.hpp]: Missing comment

File: include/mission_control.hpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        commands;  //!< Storage for current commands that
+                   //!< will be executed one by one
+
+    Position home_position;     //!< Storage for takeoff position
+    Position current_position;  //!< Current position of the drone
+
+    uint8_t current_flight_mode =
+        interfaces::msg::FlightMode::UNKNOWN;  //!< Current flight mode of the
+                                               //!< drone as published in
+                                               //!< FlightState
```
Comment (J0hn-Lawrence) [inline include/mission_control.hpp]: By "FlightState" the FlightState.msg sent on topic uav_flight_state is meant, correct? This comment is a bit unclear if the reader does not know the projects code base very well already. Maybe you could change it to "FlightState.msg" or something similar to indicate that's what it is.

File: include/mission_control.hpp [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    uint8_t current_landed_state =
+        interfaces::msg::LandedState::UNKNOWN;  //!< Current landed state of the
+                                                //!< drone as published in
+                                                //!< FlightState
```
Comment (J0hn-Lawrence) [inline include/mission_control.hpp]: see comment on line 89

File: include/mission_control.hpp [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+    bool drone_health_ok =
+        false;  //!< True if all of the health indicators in the UAVHealth.msg
+                //!< are ok, otherwise false
+
+    // Wait Time
+    static constexpr uint16_t wait_time_between_msgs =
```
Comment (J0hn-Lawrence) [inline include/mission_control.hpp]: Maybe add _ms to the variables name to indicate its unit

File: include/mission_definition_file.hpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    bool check_in_geofence(const std::array<double, 2> &point) {
+        return geofence::isIn(geofence_points_convex_hull, point);
+    }
+};
+
+/**
+ * @brief Struct representing a command.
+ *
+ * A marker can have several commands.
```
Comment (J0hn-Lawrence) [inline include/mission_definition_file.hpp:101]: By "marker" a to be detected QR Code is meant, correct? Did we explain that somewhere? The QR code detection node uses a slightly different terminology. We have to watch out in order to not confuse anyone reading the code that is not fully immersed in the project. I think it would be sufficient to add a small note about this to the README.md of this repo. What do you think?

File: include/mission_definition_file.hpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+};
+
+class MissionDefinitionReader {
+   private:
+    safety safety_settings;
+
+    /**
+     * @brief A map that associates marker names with marker commands.
+     *
+     * This map is used to store and retrieve marker commands based on their
```
Comment (J0hn-Lawrence) [inline include/mission_definition_file.hpp:115]: See comment on line 101

File: include/mission_definition_file.hpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+     *
+     * @throws std::runtime_error if the file fails to open, parse, or if the
+     * JSON content is invalid.
+     *
+     * @note This function stores the result in an internal data structure for
+     * future use, unless it is a dry run.
+     */
+    void read_file(const std::string &file_path, const bool dry_run);
+
+    safety get_safety_settings() const { return safety_settings; }
```
Comment (J0hn-Lawrence) [inline include/mission_definition_file.hpp:153]: Missing doxygen comment for this function?

File: src/basics.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
+#include "mission_control.hpp"
+
+#define ENUM_TO_STR(member) \
```
Comment (J0hn-Lawrence) [inline src/basics.cpp:18]: Missing explainatory comment. Why not move this to a header file?

File: src/basics.cpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+/**
+ * @brief Checks if the job finished successfully.
+ *
+ * This function checks if the job finished successfully and returns the result.
+ * If the job has finished successfully, the flag is reset to false.
+ *
+ * @return true if the job finished successfully, false otherwise.
+ */
+bool MissionControl::get_job_finished_successfully() {
```
Comment (J0hn-Lawrence) [inline src/basics.cpp:169]: I am unsure about the function name, as it does not indicate the reset of the job_finished flag on its own. But I am not sure about a better alternative, so if you also don't know a better alternative, we can let it slide I guess, since the comment explains it very well.

File: src/basics.cpp [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+/**
+ * @brief Retrieves the payload for a finished job.
+ *
+ * This function returns the payload for a finished job and clears the internal
+ * `job_finished_payload` variable.
+ *
+ * @return The payload for the finished job.
+ */
+nlohmann::json MissionControl::get_job_finished_payload() {
```
Comment (J0hn-Lawrence) [inline src/basics.cpp:186]: see comment on line 109, same principle

File: src/basics.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    return res;
+}
+
+/**
+ * Checks if the current mission is finished.
+ *
+ * @return true if the mission is finished, false otherwise.
+ */
+bool MissionControl::current_mission_finished() {
+    if (mission_progress >= 1.0) {
```
Comment (J0hn-Lawrence) [inline src/basics.cpp:205]: Maybe add an explanatory comment on why you are resetting this value.

File: src/basics.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    if (wait_time_finished_ok) {
+        wait_time_finished_ok = false;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/**
+ * @brief Cancels the current wait time and sets the flag to false.
```
Comment (J0hn-Lawrence) [inline src/basics.cpp]: Add name of flag to this comment please

File: src/basics.cpp [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+    // Make sure that the flag is set to false
+    wait_time_finished_ok = false;
+
+    RCLCPP_DEBUG(this->get_logger(), "MissionControl::%s: Canceled wait time",
+                 __func__);
+}
+
+/**
+ * @brief Get the string representation of the current mission state.
```
Comment (J0hn-Lawrence) [inline src/basics.cpp:271]: Maybe add somewhere that this is basically a wrapper function, as that make it less confusing with the function names.

File: src/callbacks.cpp [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+
+        mission_abort(error_msg);
+        return;
+    }
+
+    // Ignore message if sender node is not active as this might be a delayed
+    // message or a programming error in the sender node
+    if (msg.sender_id != get_active_node_id()) {
+        RCLCPP_WARN(this->get_logger(),
+                    "MissionControl::%s: Got a successfull "
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp]: Maybe add in the log that you are therefore ignoring it

File: src/callbacks.cpp [hunk 16]
```diff
@@ -0,0 +1,613 @@
...
+    if (msg.sender_id != get_active_node_id()) {
+        // Check timestamp: If it is too old, ingore message as sender might
+        // have been allowed to send at that time
+        if (timestamp_now - rclcpp::Time(msg.time_stamp) >
+            rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(
+                max_waypoint_command_msg_time_difference_ms))) {
+            // Send Error message
+            RCLCPP_ERROR(get_logger(),
+                         "MissionControl::%s: Received too "
+                         "old timestamp in waypoint command message from '%s' "
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp:196]: "from" is in the log twice, i guess by accident

File: src/callbacks.cpp [hunk 17]
```diff
@@ -0,0 +1,613 @@
...
+ *
+ * @param msg The received heartbeat message.
+ */
+void MissionControl::heartbeat_callback(const interfaces::msg::Heartbeat &msg) {
+    // Save current timestamp for later
+    rclcpp::Time timestamp_now = this->now();
+
+    // Throw away own heartbeat
+    if (msg.sender_id == get_name()) return;
+
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp:314]: Missing comment

File: src/callbacks.cpp [hunk 18]
```diff
@@ -0,0 +1,613 @@
...
+    }
+    heartbeat.tick = msg.tick;
+
+    // Check timestamp
+    if (timestamp_now - rclcpp::Time(msg.time_stamp) >
+        rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(
+            heartbeat_max_timestamp_age_ms))) {
+        RCLCPP_FATAL(get_logger(),
+                     "MissionControl::%s: Received too old "
+                     "timestamp: %s",
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp]: missing "from"

File: src/callbacks.cpp [hunk 19]
```diff
@@ -0,0 +1,613 @@
...
+    RCLCPP_DEBUG(this->get_logger(),
+                 "MissionControl::%s: Received heartbeat from: '%s', tick: %u, "
+                 "active: %d",
+                 __func__, msg.sender_id.c_str(), msg.tick, msg.active);
+}
+
+/**
+ * @brief Callback function for the heartbeat timer.
+ *
+ * This function checks if heartbeats have been received from all nodes in the
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp:425]: Is the telemetry node in the nodemap? It is not supposed to send a heartbeat and therefore should not be in there as to not trigger an error here, correct?

File: src/callbacks.cpp [hunk 20]
```diff
@@ -0,0 +1,613 @@
...
+ * node map. If a heartbeat is not received from a node, an error message is
+ * logged. If the mission is not in 'selfcheck' state, the mission will be
+ * aborted.
+ */
+void MissionControl::heartbeat_timer_callback() {
+    bool err_flag = false;
+
+    for (auto &nm : node_map) {
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp:434]: A few comments in here would not hurt

File: src/callbacks.cpp [hunk 21]
```diff
@@ -0,0 +1,613 @@
...
+
+    // Check timestamp
+    if (timestamp_now - rclcpp::Time(msg.time_stamp) >
+        rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(
+            max_position_msg_time_difference_ms))) {
+        // Reset current_position as we can no longer reliabely know where we
+        // are
+        current_position = Position();
+
+        if (this->get_active()) {
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp:545]: Missing comment

File: src/callbacks.cpp [hunk 22]
```diff
@@ -0,0 +1,613 @@
...
+    const interfaces::msg::MissionProgress &msg) {
+    // Store current timestamp for later
+    rclcpp::Time timestamp_now = this->now();
+
+    RCLCPP_DEBUG(this->get_logger(),
+                 "MissionControl::%s: Received mission "
+                 "progress from '%s': progress: %f / 1.0",
+                 __func__, msg.sender_id.c_str(), msg.progress);
+
+    // Check timestamp
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp]: maybe add in the comment what you are checking the timestamp for -> too old ,this is appears multiple times and should be changed for all instances if you decide to change it

File: src/mission_definition_file.cpp
Comment (J0hn-Lawrence) [inline src/mission_definition_file.cpp:1]: Why use printf here? I assume because you are also using this code in the "static" analysis of the mission definition which is standalone?

File: src/modes.cpp [hunk 23]
```diff
@@ -0,0 +1,468 @@
...
+        set_mission_state(selfcheck);
+    }
+}
+
+/**
+ * @brief Initiates the takeoff procedure.
+ *
+ * This function is responsible for initiating the takeoff procedure of the
+ * drone. It performs several checks to ensure that the drone is in the correct
+ * state and has is currently inside the geofence before sending the takeoff
```
Comment (J0hn-Lawrence) [inline src/modes.cpp]: "has" should be removed here

File: src/modes.cpp [hunk 24]
```diff
@@ -0,0 +1,468 @@
...
+    if (get_state_first_loop()) {
+        RCLCPP_INFO(this->get_logger(), "MissionControl::%s: Takeoff initated",
+                    __func__);
+
+        // Activate Mission Control
+        send_control_json(this->get_name(), true, {});
+        this->activate();
+        set_active_node_id(this->get_name());
+
+        // Check that drone is in 'READY' state and landed state is 'ON_GROUND'
```
Comment (J0hn-Lawrence) [inline src/modes.cpp]: mode and state mean the same thing here, right?

**Review by J0hn-Lawrence — CHANGES_REQUESTED** [review]
Could not yet take a deeper look into modes.cpp due to time, but I think the feedback is still valuable so providing it already now. See comments, overall very nice progress!

File: assets/mission_test.json
Comment (Philipp66904) [reply assets/mission_test.json:1]: Yes, I just wanted to use real values from Google Maps so I can check if the geofencing works by using Values that I know are inside or outside the geofence. Therefore, the real site just seemed logical ;)

File: include/mission_control.hpp (hunk 1 above)
Comment (Philipp66904) [reply include/mission_control.hpp:70]: It's always the last one.
This is meant for the appropriate state to poll if the job finished (stored in the flag) and then get the last results from there (e. g. the decoded marker). This way, not the callback but the current mode function (= triggered periodically by Event Loop) has to decode this data, where all the functionality is grouped together this way and not spread around into several callbacks.

File: include/mission_control.hpp (hunk 2 above)
Comment (Philipp66904) [reply include/mission_control.hpp]: Added comment

File: include/mission_control.hpp (hunk 3 above)
Comment (Philipp66904) [reply include/mission_control.hpp]: Added that one and also the other similar cases

File: include/mission_control.hpp (hunk 5 above)
Comment (Philipp66904) [reply include/mission_control.hpp]: Added

File: include/mission_definition_file.hpp (hunk 6 above)
Comment (Philipp66904) [reply include/mission_definition_file.hpp:101]: I added a `Terminology` chapter to the documentation: https://docs.google.com/document/d/1BV6CUhAO0_3A8xje5DsFQydK13K3FuRfkojx6XGCMnk/edit#heading=h.l87p31sq2caw

File: include/mission_definition_file.hpp (hunk 8 above)
Comment (Philipp66904) [reply include/mission_definition_file.hpp:153]: I usually don't comment simple getters like this, but I added it here because it is a library and it doesn't really hurt here.

File: src/basics.cpp (hunk 9 above)
Comment (Philipp66904) [reply src/basics.cpp:18]: I added a comment.
It is located in this file, because it is only used here.

File: src/basics.cpp (hunk 12 above)
Comment (Philipp66904) [reply src/basics.cpp:205]: Done

File: src/basics.cpp (hunk 13 above)
Comment (Philipp66904) [reply src/basics.cpp]: Done

File: src/basics.cpp (hunk 14 above)
Comment (Philipp66904) [reply src/basics.cpp:271]: Done

File: src/callbacks.cpp (hunk 15 above)
Comment (Philipp66904) [reply src/callbacks.cpp]: Done

File: src/callbacks.cpp (hunk 16 above)
Comment (Philipp66904) [reply src/callbacks.cpp:196]: I will just blame it on the autoformatter 😉

File: src/callbacks.cpp (hunk 17 above)
Comment (Philipp66904) [reply src/callbacks.cpp:314]: Done

File: src/callbacks.cpp (hunk 18 above)
Comment (Philipp66904) [reply src/callbacks.cpp]: Done

File: src/callbacks.cpp (hunk 19 above)
Comment (Philipp66904) [reply src/callbacks.cpp:425]: The node map is initialized in the constructor (`mission_control.cpp`) of Mission Control.
And no, the telemetry node is no longer in there.

File: src/callbacks.cpp (hunk 20 above)
Comment (Philipp66904) [reply src/callbacks.cpp:434]: Done

File: src/callbacks.cpp (hunk 21 above)
Comment (Philipp66904) [reply src/callbacks.cpp:545]: Added

File: src/callbacks.cpp (hunk 22 above)
Comment (Philipp66904) [reply src/callbacks.cpp]: Changed for all occurrences

File: src/mission_definition_file.cpp
Comment (Philipp66904) [reply src/mission_definition_file.cpp:1]: Yes, that's correct. I want it to be free of any ROS dependencies and that's why I'm using printf here.
The final error message will be packed into the exception and thereby be logged into our ROS network.
The debugging of the file should then happen with the standalone program (or in the best case before putting it on the drone).

File: src/modes.cpp (hunk 23 above)
Comment (Philipp66904) [reply src/modes.cpp]: Fixed

File: src/modes.cpp (hunk 24 above)
Comment (Philipp66904) [reply src/modes.cpp]: Yes

**Comment by Philipp66904** [comment]
Thanks, I already implemented your comments 😉

**Comment by J0hn-Lawrence** [comment]
> Thanks, I already implemented your comments 😉

Thank you, looks good! Tag me when you need another approval for merging or review.

**Comment by J0hn-Lawrence** [comment]
@Philipp66904 please tag me once you want me to review

**Comment by Philipp66904** [comment]
[Colcon Build, Test, Coverage and Autoformat](https://github.com/THI-Drone/mission_control_package/actions/workflows/workflow.yml)

**Comment by Philipp66904** [comment]
I think we should push this out so that V&V can test it.
The tests are still work in progress and I will open a new PR for them.

@J0hn-Lawrence 
@J0las

**Comment by J0hn-Lawrence** [comment]
Do you already want me to review @Philipp66904 ?

**Comment by Philipp66904** [comment]
@J0hn-Lawrence
Yes, as otherwise we don't have a working version in `dev`

**Comment by J0hn-Lawrence** [comment]
> @J0hn-Lawrence Yes, as otherwise we don't have a working version in `dev`

Will do my best to finish the review today.

File: test/mission_file_reader/test_assets/mdf_too_less_geofence_points.json
Comment (J0hn-Lawrence) [inline test/mission_file_reader/test_assets/mdf_too_less_geofence_points.json:1]: file name is grammatically incorrect, should be too little instead of too less, leave it if you don't care, just had to note it

File: src/callbacks.cpp [hunk 25]
```diff
@@ -569,32 +690,6 @@ void MissionControl::health_callback(const interfaces::msg::UAVHealth &msg) {
                       msg.is_magnetometer_calibration_ok &&
                       msg.is_local_position_ok && msg.is_global_position_ok &&
                       msg.is_home_position_ok && msg.is_armable;
-
-    // Check that drone health is still good
```
Comment (J0hn-Lawrence) [inline src/callbacks.cpp]: Why did you remove this section? I cannot immediately spot if you maybe put this elsewhere.

**Review by J0hn-Lawrence — COMMENTED** [review]
Very nice progress, thank you for all of the work! Some minor comments

File: test/mission_file_reader/test_assets/mdf_too_less_geofence_points.json
Comment (Philipp66904) [reply test/mission_file_reader/test_assets/mdf_too_less_geofence_points.json:1]: You're right, but I don't feel like changing it 🤣

File: src/callbacks.cpp (hunk 25 above)
Comment (Philipp66904) [reply src/callbacks.cpp]: I removed this section because the fcc bridge takes care of the drone health and I only use this check before I arm.

@J0las and I decided to split the responsibilities: @J0las cares about the drone and its health, while I care about the mission and the other nodes.
I thought that duplicating the effort just leads to problems.

**Comment by Philipp66904** [comment]
I would like to merge this into dev before the test runs tomorrow.
Please add your approval @J0hn-Lawrence @J0las

**Review by J0hn-Lawrence — APPROVED** [review]
Approved

**Review by J0las — APPROVED** [review]
Philipp ist schuld wenn das net funktioniert.
